crates/ruff_python_parser/src/lexer.rs

```diff
@@ -133,13 +133,24 @@ impl<'src> Lexer<'src> {
         std::mem::take(&mut self.current_value)
     }
 
+    /// Helper function to push the given error, updating the current range with the error location
+    /// and return the [`TokenKind::Unknown`] token.
+    fn push_error(&mut self, error: LexicalError) -> TokenKind {
+        self.current_range = error.location();
+        self.errors.push(error);
+        TokenKind::Unknown
+    }
+
     /// Lex the next token.
     pub fn next_token(&mut self) -> TokenKind {
         self.cursor.start_token();
         self.current_value = TokenValue::None;
         self.current_flags = TokenFlags::empty();
         self.current_kind = self.lex_token();
-        self.current_range = self.token_range();
+        // For `Unknown` token, the `push_error` method updates the current range.
+        if !matches!(self.current_kind, TokenKind::Unknown) {
+            self.current_range = self.token_range();
+        }
```

Comment:
It's a bit unfortunate that we now have this branch in the very hot `next_token` function for the very rare case of an unknown token. But I don't see a better way of solving this that doesn't require a lot of repetition on the token range.

Reply:
Yeah, I agree. I thought of doing something like what you've suggested (computing the token range at the place where it's emitted) but wanted to see if this actually impacts the performance. It doesn't seem to be which is why I'm fine moving ahead with this.